packages/react-pages/src/node/dynamic-modules/PagesData.ts

```diff
@@ -0,0 +1,289 @@
+import deepEqual from 'fast-deep-equal'
+import { File } from './PageStrategy'
+import type { ScheduleUpdate } from './UpdateBuffer'
+
+export class PagesDataKeeper {
+  private pages: PagesDataInternal = {}
+
+  /**
+   * If the page page does not come from local file,
+   * then we create this api to track the source of page data.
+   * When a file is unlinked,
+   * the data associated with it will automatically get deleted.
+   */
+  createAPIForSourceFile(
+    sourceFile: File,
+    scheduleUpdate: ScheduleUpdate
+  ): HanlderAPI {
+    return {
+      getRuntimeData: (
+        pageId: string
+      ): {
+        [key: string]: string
+      } =>
+        this.createMutableProxy(pageId, 'runtime', sourceFile, scheduleUpdate),
+      getStaticData: (
+        pageId: string
+      ): {
+        [key: string]: any
+      } =>
+        this.createMutableProxy(pageId, 'static', sourceFile, scheduleUpdate),
+    }
+  }
+
+  /**
+   * If the page page does not come from local file (.e.g from remote database),
+   * then we can't track the source of page data.
+   */
+  createAPIForCustomSource(scheduleUpdate: ScheduleUpdate): HanlderAPI {
+    return {
+      getRuntimeData: (
+        pageId: string
+      ): {
+        [key: string]: string
+      } => this.createMutableProxy(pageId, 'runtime', null, scheduleUpdate),
+      getStaticData: (
+        pageId: string
+      ): {
+        [key: string]: any
+      } => this.createMutableProxy(pageId, 'static', null, scheduleUpdate),
+    }
+  }
+
+  private createMutableProxy(
+    pageId: string,
+    dataType: 'runtime' | 'static',
+    sourceFile: File | null,
+    scheduleUpdate: ScheduleUpdate
+  ): {
+    [key: string]: any
+  } {
+    const _this = this
+    const dataObj = this.getDataObj(pageId, dataType)
+    return new Proxy(dataObj, {
+      set(target, prop: string, value) {
+        if (value === undefined) {
+          // undefined means 'delete'
+          const existValue = dataObj[prop]
+          if (existValue) {
+            existValue.unBind(scheduleUpdate)
+          }
+          return true
+        }
+        new Association(
+          _this,
+          pageId,
+          dataType,
+          prop,
+          value,
+          sourceFile,
+          scheduleUpdate
+        )
+        return true
+      },
+      get(target, prop: string) {
+        const association = target[prop]
+        if (association === undefined) return undefined
+        return association.value
+      },
+    })
+  }
+
+  toPagesData(): PagesData {
+    const result: PagesData = {}
+    Object.entries(this.pages).forEach(([pageId, page]) => {
```

Comment:
Recreating the pages manifest on every change seems like it could slow down the feedback loop for sites with hundreds of pages, but I guess we can cross that bridge if/when we get there.

Reply:
Updates can be batched if they happens closely. We could increase the debounce time if/when we have performance problem.